app def missing.

RubelMozumder opened the pull request:
@budschi, @ka-sarthak  and @sherjeelshabih , Now the nexus parts are available in `EntryArchive.json`. But it needs the MR in [gtilab](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/merge_requests/1971) merged to the develop branch.

The next task is to map the full data path from archive to nexus and vice-versa. Once @ka-sarthak comes from holidays we can map all the data and merge it.

ka-sarthak commented:
from our meeting today:
- Entry type of ELNXRayDiffraction should not be modified
- ELNXRayDiffraction should have a **nexus sub-section**
- additional .nxs file can be created
- additional entry based on the .nxs file should be avoided

ka-sarthak commented:
@RubelMozumder 
After discussing with @hampusnasstrom, we think it's okay to have an entry archive for the corresponding .nxs file. It indexes the .nxs file for the nomad search.

The important thing is that data should not be duplicated. But I think this is already avoided because both the nexus sub-section and the entry archive based on .nxs file do not contain the data, but references to the quantities in .nxs file. Right?

As such, the only thing remaining to resolve is that the entry type of ELNXRayDiffraction should not be modified.

RubelMozumder commented:
> @RubelMozumder FYI I rebased the `write-nexus-section` branch, but now your branch has some merge conflicts

I can fix it.

RubelMozumder commented:
> @RubelMozumder some comments from our meeting today:
> 
> * Entry type of ELNXRayDiffraction should not be modified
> * ELNXRayDiffraction should have a **nexus sub-section**
> * additional .nxs file can be created
> * additional entry based on the .nxs file should be avoided

All requested features are implemented, here. A short overview:
1. Nexus hdf5 file can be created as an entry or a simple file using boolean arg: `nxs_as_entry`.
2. The schema type or entry_type `ELNXRayDiffraction` is not overwritten anymore.

I separated the related code into a file `nx.py` in the interest of code readability and maintainability.

RubelMozumder commented:
@ka-sarthak, I think it should be ready to merge, now.

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        try:
+            template['/ENTRY[entry]/experiment_result/intensity'] = archive.data.results[0].intensity.magnitude
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_result/two_theta'] = archive.data.results[0].two_theta.magnitude
+            template['/ENTRY[entry]/experiment_result/two_theta/@units'] = archive.data.results[0].two_theta.units.__str__()
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_result/omega'] = archive.data.results[0].omega.magnitude
+            template['/ENTRY[entry]/experiment_result/omega/@units'] = archive.data.results[0].omega.units.__str__()
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_result/chi'] = archive.data.results[0].chi.magnitude
+            template['/ENTRY[entry]/experiment_result/chi/@units'] = archive.data.results[0].chi.units.__str__()
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_result/phi'] = archive.data.results[0].phi.magnitude
+            template['/ENTRY[entry]/experiment_result/phi/@units'] = archive.data.results[0].phi.units.__str__()
+        except AttributeError as e:
+            pass
+
+        try:
+            template[
+                '/ENTRY[entry]/INSTRUMENT[instrument]/DETECTOR[detector]/scan_axis'
+            ] = archive.data.results[0].scan_axis
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_config/count_time'] = archive.data.results[0].count_time.magnitude
+        except AttributeError as e:
+            pass
```
```suggestion
```

These quantities are common for `rsm` and `line` scans. You can take it out of the if clause,

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        try:
+            template['/ENTRY[entry]/experiment_result/intensity'] = archive.data.results[0].intensity.magnitude
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_result/two_theta'] = archive.data.results[0].two_theta.magnitude
+            template['/ENTRY[entry]/experiment_result/two_theta/@units'] = archive.data.results[0].two_theta.units.__str__()
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_result/omega'] = archive.data.results[0].omega.magnitude
+            template['/ENTRY[entry]/experiment_result/omega/@units'] = archive.data.results[0].omega.units.__str__()
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_result/chi'] = archive.data.results[0].chi.magnitude
+            template['/ENTRY[entry]/experiment_result/chi/@units'] = archive.data.results[0].chi.units.__str__()
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_result/phi'] = archive.data.results[0].phi.magnitude
+            template['/ENTRY[entry]/experiment_result/phi/@units'] = archive.data.results[0].phi.units.__str__()
+        except AttributeError as e:
+            pass
+
+        try:
+            template[
+                '/ENTRY[entry]/INSTRUMENT[instrument]/DETECTOR[detector]/scan_axis'
+            ] = archive.data.results[0].scan_axis
+        except AttributeError as e:
+            pass
+
+        try:
+            template['/ENTRY[entry]/experiment_config/count_time'] = archive.data.results[0].count_time.magnitude
+        except AttributeError as e:
+            pass
```
```suggestion
    try:
        template['/ENTRY[entry]/experiment_result/intensity'] = archive.data.results[0].intensity.magnitude
    except AttributeError:
        pass
    
    try:
        template['/ENTRY[entry]/experiment_result/two_theta'] = archive.data.results[0].two_theta.magnitude
        template['/ENTRY[entry]/experiment_result/two_theta/@units'] = archive.data.results[0].two_theta.units.__str__()
    except AttributeError:
        pass
    
    try:
        template['/ENTRY[entry]/experiment_result/omega'] = archive.data.results[0].omega.magnitude
        template['/ENTRY[entry]/experiment_result/omega/@units'] = archive.data.results[0].omega.units.__str__()
    except AttributeError:
        pass
    
    try:
        template['/ENTRY[entry]/experiment_result/chi'] = archive.data.results[0].chi.magnitude
        template['/ENTRY[entry]/experiment_result/chi/@units'] = archive.data.results[0].chi.units.__str__()
    except AttributeError:
        pass
    
    try:
        template['/ENTRY[entry]/experiment_result/phi'] = archive.data.results[0].phi.magnitude
        template['/ENTRY[entry]/experiment_result/phi/@units'] = archive.data.results[0].phi.units.__str__()
    except AttributeError:
        pass
    
    try:
        template[
            '/ENTRY[entry]/INSTRUMENT[instrument]/DETECTOR[detector]/scan_axis'
        ] = archive.data.results[0].scan_axis
    except AttributeError:
        pass
    
    try:
        template['/ENTRY[entry]/experiment_config/count_time'] = archive.data.results[0].count_time.magnitude
    except AttributeError:
        pass
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        try:
+            template['/ENTRY[entry]/experiment_result/q_parallel'] = archive.data.results[0].q_parallel,
+            template['/ENTRY[entry]/experiment_result/q_parallel/@units'] = archive.data.results[0].q_parallel.units.__str__()
+        except AttributeError as e:
```
```suggestion
        except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        try:
+            template['/ENTRY[entry]/experiment_result/q_perpendicular'] = archive.data.results[0].q_perpendicular.magnitude
+            template['/ENTRY[entry]/experiment_result/q_perpendicular/@units'] = archive.data.results[0].q_perpendicular.units.__str__()
+        except AttributeError as e:
```
```suggestion
        except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        try:
+            template['/ENTRY[entry]/experiment_result/q_norm'] = archive.data.results[0].q_norm.magnitude
+            template['/ENTRY[entry]/experiment_result/q_norm/@units'] = archive.data.results[0].q_norm.units.__str__()
+        except AttributeError as e:
```
```suggestion
        except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        template[
+            '/ENTRY[entry]/INSTRUMENT[instrument]/SOURCE[source]/xray_tube_material'
+        ] = archive.data.xrd_settings.source.xray_tube_material
+    except AttributeError as e:
```
```suggestion
    except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        template[
+            '/ENTRY[entry]/INSTRUMENT[instrument]/SOURCE[source]/xray_tube_current/@units'
+        ] = archive.data.xrd_settings.source.xray_tube_current.units.__str__()
+    except AttributeError as e:
```
```suggestion
    except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        template[
+            '/ENTRY[entry]/INSTRUMENT[instrument]/SOURCE[source]/xray_tube_voltage/@units'
+        ] = archive.data.xrd_settings.source.xray_tube_voltage.units.__str__()
+    except AttributeError as e:
```
```suggestion
    except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        template[
+            '/ENTRY[entry]/INSTRUMENT[instrument]/SOURCE[source]/k_alpha_one/@units'
+        ] = archive.data.xrd_settings.source.kalpha_one.units.__str__()
+    except AttributeError as e:
```
```suggestion
    except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        template[
+            '/ENTRY[entry]/INSTRUMENT[instrument]/SOURCE[source]/k_alpha_two/@units'
+        ] = archive.data.xrd_settings.source.kalpha_two.units.__str__()
+    except AttributeError as e:
```
```suggestion
    except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        template[
+            '/ENTRY[entry]/INSTRUMENT[instrument]/SOURCE[source]/ratio_k_alphatwo_k_alphaone'
+        ] = archive.data.xrd_settings.source.ratio_kalphatwo_kalphaone
+    except AttributeError as e:
```
```suggestion
    except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        template[
+            '/ENTRY[entry]/INSTRUMENT[instrument]/SOURCE[source]/kbeta/@units'
+        ] = archive.data.xrd_settings.source.kbeta.units.__str__()
+    except AttributeError as e:
```
```suggestion
    except AttributeError:
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+
+# source_peak_wavelength
+def connect_concepts(template, archive: 'EntryArchive', scan_type: str):
+    """Connect the concepts between ELNXrayDiffraction and NXxrd_pan schema."""
```
Please add a more explicit doc string including args and returns.

ka-sarthak reviewed (changes requested):
Looks good! I have added some refactoring suggestions. Once you made those changes, please add for a review again.

Also, can you run `ruff format src/nomad_measurements/xrd/nx.py`

RubelMozumder commented on src/nomad_measurements/xrd/schema.py:
```diff
@@ -80,19 +80,17 @@
-    two_theta: 'pint.Quantity' = None,
-) -> tuple['pint.Quantity', 'pint.Quantity']:
-    """
+        wavelength: 'pint.Quantity',
+        q: 'pint.Quantity'=None,
+        two_theta: 'pint.Quantity'=None
+    ) -> tuple['pint.Quantity', 'pint.Quantity']:
+    '''
```
Are you using any code formater like `ruff` or `black` if yes what is the which one and what is the version of that. It is cumbersome doing manually as I always use `ruff` format of code on save which reduces effort in code maintaining. Because in the other PR you will see same issue again.

ka-sarthak replied:
To avoid running the ruff formatting, you can set this setting in `.vscode/setting.json`:
```json
"editor.formatOnSave": false
```

ka-sarthak replied:
yeah, totally, I also rely on it heavily. But at least for this PR, it's better to turn it off.

RubelMozumder commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+        on_temp_file=generate_nexus_file,
+        nxs_as_entry=nxs_as_entry
+    )
+    archive.metadata.entry_type = entry_type
```
It is not the functionality of pynxtools. The ´entry_type` is picked by nomad. Nomad does check it the entry type already exists or not rather write the entry name directly. 
`NXxrd_pad` is also an entry type as the process take place under the `ELNxrayDiffraction` and nomad overwrites the `entry_name`. Wherever I fix it in the normalize process the `entry_name` gets back.

ka-sarthak replied:
But the entry_type is already present: `archive.metadata.entry_type = "ELNXRayDiffraction"`, it get's modified in  the function from pynxtools. Which is why you are reverting it back. What I mean is that this should be resolved inside the pynxtools: why is the entry type being replaced?

ka-sarthak replied:
Maybe it's worth connecting with someone in AreaD to resolve this. Did you check if there is any other metadata also being overwritten?

RubelMozumder replied:
What is the problem with this approach, the entry name has been written as hard code, So the entry name has been stored and back to the `entry_type`. I do not see anything wrong with that. End of the day schema.py is the authority of the `entry_type`.

ka-sarthak replied:
No Rubel, it doesn't seem right. What you offer is a quick solution which works for now. But if the m_def of an entry is being overwritten by a function, we need to fix it in that function. More importantly, we need to know why it is being overwritten in the first place, to be sure that other metadata is not being overwritten.

We can ask someone in Area B or D who knows about this code.

RubelMozumder commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,264 @@
+    """Connect the concepts between ELNXrayDiffraction and NXxrd_pan schema."""
+
+    # Genneral concepts
+    try:
+        template['/ENTRY[entry]/definition'] = 'NXxrd_pan'
+    except AttributeError as e:
+        pass
```
I would remove it, I just used it so that we follow the same code style, though we do not access the `e` obj anywhere.

ka-sarthak commented on pyproject.toml:
```diff
@@ -26,7 +26,9 @@ classifiers = [
-    "fairmat-readers-xrd~=0.0.3"
+    "fairmat-readers-xrd~=0.0.3",
+    "pynxtools>=0.6.0",
+    "pynxtools-xrd>=0.0.2"
```
I see, but we can still add the same dependency in the `pynxtools` repo. Since nomad-measurements does not directly use pynxtools-xrd, I don't see a reason to make it a dep here.

RubelMozumder replied:
We have it there, but still developer (at least you guys) needs to know which version is needed while debugging. End of the day you will get `pynxtools-xrd` via `pynxtools` it is just the wording of the package nothing else. However, I will fix it.

ka-sarthak commented:
I tried running the branch in my local nomad installation. For some reason, the entry associated with the .nxs file is not being generated. Didn't you implement this? Or do I need to add nexus parsers for it to work?

RubelMozumder commented on nomad.yaml:
```diff
@@ -5,6 +5,9 @@ normalize:
 plugins:
   include: 
     - parsers/nomad_measurements/xrd
+    - pynxtools.nomad.entrypoints:nexus_data_converter
+    - pynxtools.nomad.entrypoints:nexus_parser
+    - pynxtools.nomad.entrypoints:nexus_schema
```
Keep it in mind without these plugins of pynxtools `nomad_measurement` would not work.

ka-sarthak commented on src/nomad_measurements/xrd/schema.py:
```diff
@@ -1000,7 +970,9 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger'):
             return
-        self.write_nx_section_and_create_file(archive, logger)
+        write_nx_section_and_create_file(
+            archive, logger, self.generate_nexus_file, nxs_as_entry=True
```
`self.generate_nexus_file` does not seem to work as intended. Even if it's set as false, the nexus file is still generated

RubelMozumder replied:
Sorry it is a wrong logic. I fix it. Can you try it.

ka-sarthak replied:
looks good! can you update the the argument descriptions in docstring of `write_nx_section_and_create_file`?

ka-sarthak commented on src/nomad_measurements/xrd/schema.py:
```diff
@@ -24,6 +24,7 @@
 import numpy as np
 import plotly.express as px
 from scipy.interpolate import griddata
+from traitlets import default
```
I don't think we need this, or?

ka-sarthak replied:
```suggestion
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,222 @@
+
+    try:
+        template['/ENTRY[entry]/experiment_result/two_theta'] = archive.data.results[0].two_theta.magnitude
+        template['/ENTRY[entry]/experiment_result/two_theta/@units'] = archive.data.results[0].two_theta.units.__str__()
```
I just ran the linter and it suggests using `str()` method instead of the dunder call `.__str__()`

here's some docu for this https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/unnecessary-dunder-call.html

```suggestion
        template['/ENTRY[entry]/experiment_result/two_theta/@units'] = str(archive.data.results[0].two_theta.units)
```

ka-sarthak replied:
It would be nice to stick with the lining suggestions. Please make similar changes in other places too.

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,222 @@
+    from nomad.datamodel.datamodel import EntryArchive
+
+
+# source_peak_wavelength
```
seems like a dangling comment, can get rid of it

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,222 @@
+
+# source_peak_wavelength
+def connect_concepts(template, archive: 'EntryArchive', scan_type: str):
+    """Connect the concepts between ELNXrayDiffraction and NXxrd_pan schema.
```
```suggestion
    """
    Connect the concepts between `ELNXRayDiffraction` and `NXxrd_pan` schema.
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,222 @@
+        scan_type (str): Name of the scan type such as line and RSM.
+    """
+
+    # Genneral concepts
```
```suggestion
    # General concepts
```

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,222 @@
+        logger=logger,
+        output_file_path=nexus_output,
+    )
+    archive.metadata.entry_type = entry_type
```
```suggestion
    # TODO: check why is the entry_type overwritten;
    # Is there other metadata being overwritten?  
    archive.metadata.entry_type = entry_type
```

ka-sarthak commented:
I have added some comments. Once you resolve it, please do the following things:
- run the ruff autoformatter on nx.py file (because this is newly added code, better to be formatted)
- remove all the formatting from schema.py (because this is old code, and formatting will be handled elsewhere)

One issue that remains is: 
m_def of an entry is being overwritten by `populate_nexus_subsection`. Your current fix to re-overwrite it again works for now, but we need to be sure that other metadata is not being altered. As suggested, we can add a TODO there to close this PR and return to it later.

But otherwise, looks great! Thanks for the amazing work!

lukaspie commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,256 @@
+    # General concepts
+    template['/ENTRY[entry]/definition'] = 'NXxrd_pan'
+
+    try:
```
@RubelMozumder maybe I am missing somethin (correct me if I'm wrong), but wouldn't it be good to use a mapping here instead of testing each concept one by one?

So something like
```
concept_mapping: Dict[str, Any] = {
    "/ENTRY[entry]/method": archive.data.method,
    "/ENTRY[entry]/measurement_type": archive.data.diffraction_method_name,
    "/ENTRY[entry]/experiment_result/intensity": archive.data.results[0].intensity.magnitude,
    ...
}
for nxs_concept, archive_path in concept_mapping.items():
    try:
        template[nxs_concept] = archive_path
    except AttributeError:
        pass
````
Wouldn't that save a lot of `try-except` repetitions? I think it is also good to have a such mapping in one place, so one can keep track of it.

RubelMozumder replied:
Hi @lukaspie, the code on the right side will be executed immediately and raise an error (if any error is encountered) before going to the `try-except` block in for loop.

lukaspie replied:
Hm, I see, that's indeed unfortunate. One could probably do something more complicated, like this:
```python
from functools import reduce
import operator

def safe_getattr(obj, attr_path, default=None):
    """Safely navigate nested attributes and indices based on a dot-separated path."""
    try:
        for attr in attr_path.split('.'):
            if '[' in attr and ']' in attr:  # Handle list indexing like results[0]
                attr_name, index = attr[:-1].split('[')
                obj = getattr(obj, attr_name)  # Get the list
                obj = obj[int(index)]  # Access the specific index
            else:
                obj = getattr(obj, attr)  # Regular attribute access
        return obj
    except (AttributeError, IndexError, KeyError, TypeError):
        return default

concept_mapping: Dict[str, str] = {
    "/ENTRY[entry]/method": "method",
    "/ENTRY[entry]/measurement_type": "diffraction_method_name",
    "/ENTRY[entry]/experiment_result/intensity": "results[0].intensity.magnitude",
    # Add more mappings as needed
}

for concept, archive_attr in concept_mapping.items():
    try:
        # Check if archive.data exists first
        data = getattr(archive, 'data', None)
        if data is not None:
            archive_path = safe_getattr(data, archive_attr)
            if archive_path is not None:
                template[concept] = archive_path
    except (AttributeError, IndexError):
        pass
```
But you can also keep it like it is for now. We could maybe come back to this later if we want more of such mappings and want to organize this a bit better.

RubelMozumder commented:
> I have added some comments. Once you resolve it, please do the following things:
> 
> * run the ruff autoformatter on nx.py file (because this is newly added code, better to be formatted)
> * remove all the formatting from schema.py (because this is old code, and formatting will be handled elsewhere)
> 
> One issue that remains is: m_def of an entry is being overwritten by `populate_nexus_subsection`. Your current fix to re-overwrite it again works for now, but we need to be sure that other metadata is not being altered. As suggested, we can add a TODO there to close this PR and return to it later.
> 
> But otherwise, looks great! Thanks for the amazing work!

@ka-sarthak, I removed the `entry_type` overwriting part. From my side it is ok. Would you please test it to see if it is okay with you?

please, keep in mind to test the changes in pynxtools you have to try from the `master` branch from https://github.com/FAIRmat-NFDI/pynxtools. A new release will be available by tomorrow.

ka-sarthak commented:
@RubelMozumder 
As suggested by @aalbino2 , I just rebased the `write-nexus-section` on `101-update-package-structure` to include the migration to the new plugin mechanism. I also rebased the current PR branch on `write-nexus-section` to include those changes here. 

now if you want to test the plugin in your local installation, make sure to use the new plugin mechanism to load the plugin.

ka-sarthak commented:
@RubelMozumder  `entry_type` of ELNXRayDiffraction is getting modified when the nexus section is populated. it seems that the changes you made in pynxtools does not fix this issue. can you take another look?

ka-sarthak commented:
or maybe it's that pynxtools version need to be updated in pyproject

aalbino2 commented:
thanks @ka-sarthak 

Hey @RubelMozumder we can look at the rebase together if you need, let me know

ka-sarthak commented on src/nomad_measurements/xrd/schema.py:
```diff
@@ -1014,8 +977,12 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger'):
+        if self.generate_nexus_file and self.data_file is not None:
+            write_nx_section_and_create_file(archive, logger, scan_type=scan_type)
+
+            self.figures = self.results[0].generate_plots(archive, logger)
```
```suggestion
        self.figures = self.results[0].generate_plots(archive, logger)
```

The plotting of figures should not be inside the if block.

ka-sarthak commented on pyproject.toml:
```diff
@@ -32,7 +32,7 @@ dependencies = [
     "xmltodict==0.13.0",
     "fairmat-readers-xrd~=0.0.3",
-    "pynxtools", 
+    "pynxtools>=0.6.0",
```
```suggestion
    "pynxtools>=0.6.1",
```

Let's update the pynxtools to the version where you made the changes in `populate_nexus_subsection`

ka-sarthak commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,184 @@
+    template['/ENTRY[entry]/@default'] = 'experiment_result'
+    template['/ENTRY[entry]/experiment_result/@signal'] = 'intensity'
+    template['/ENTRY[entry]/experiment_result/@axes'] = 'two_theta'
+    template['/ENTRY[entry]/q_data/q'] = {
+        'link': '/ENTRY[entry]/experiment_result/q_norm'
+    }
+    template['/ENTRY[entry]/q_data/intensity'] = {
+        'link': '/ENTRY[entry]/experiment_result/intensity'
+    }
+    template['/ENTRY[entry]/q_data/q_parallel'] = {
+        'link': '/ENTRY[entry]/experiment_result/q_parallel'
+    }
+    template['/ENTRY[entry]/q_data/q_perpendicular'] = {
+        'link': '/ENTRY[entry]/experiment_result/q_perpendicular'
+    }
```
I wonder why you are using "link" for only these quantities and not for others. What's the meaning of "link" in templates?

RubelMozumder replied:
This is how the `NXxrd_pan` app def has been defined which is similar to the reference, avoiding the duplication of array data.

ka-sarthak replied:
I see. Then I would suggest doing the same for other array quantities like two_theta, omega, etc. For now, we can stick with the current version.

ka-sarthak reviewed (changes requested):
The refactored code in nx.py is very concise. Good work there! 
Added comments to bump up `pynxtools` in `pyproject.toml` and fix an error, where `generate_plots` has been moved into the `if` block.

RubelMozumder commented on src/nomad_measurements/xrd/nx.py:
```diff
@@ -0,0 +1,184 @@
+                    finally:
+                        if value is not None:
+                            template[sub_key] = (
+                                str(value)
+                                if archive_concept.endswith('units')
+                                else value
```
```suggestion
                                str(value)
                                if sub_key.endswith('units')
                                else value
```

ka-sarthak reviewed (approved):
The tests are now working fine. Awesome work! I am approving it.
We will need to add some tests at some point to test if the .nxs file generated has the correct data.
